libs/MeshKernel/include/MeshKernel/RangeCheck.hpp

```diff
@@ -0,0 +1,501 @@
+// P.O. Box 177
+// 2600 MH Delft, The Netherlands
+//
+// All indications and logos of, and references to, "Delft3D" and "Deltares"
+// are registered trademarks of Stichting Deltares, and remain the property of
+// Stichting Deltares. All rights reserved.
+//
+//------------------------------------------------------------------------------
+
+#pragma once
+
+#include "MeshKernel/Exceptions.hpp"
+#include "MeshKernel/Formatting.hpp"
+
+#include <algorithm>
+#include <concepts>
+#include <functional>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <utility>
+
+namespace meshkernel
+{
+    namespace range_check
+    {
+        /// @brief Defines the checkable types: floating point types and inetgral types except for bool and char.
+        template <typename T>
+        concept RangeCheckableType = std::floating_point<T> ||
+                                     (std::integral<T> &&
+                                      !std::same_as<T, bool> &&
+                                      !std::same_as<T, char> &&
+                                      !std::same_as<T, char8_t> &&
+                                      !std::same_as<T, char16_t> &&
+                                      !std::same_as<T, char32_t> &&
+                                      !std::same_as<T, wchar_t>);
+
+        /// @brief  Keys of performed comparison
+        enum class Comparison
+        {
+            Equal,                       ///< Equal to
+            NotEqual,                    ///< Not equal to
+            Greater,                     ///< Greater than
+            GreaterEqual,                ///< Greater than or equal to
+            Less,                        ///< Less than
+            LessEqual,                   ///< Less than or equal to
+            InClosedInterval,            ///< In closed interval
+            InOpenInterval,              ///< In open open interval
+            InSemiOpenFromAboveInterval, ///< In semi-open from above interval
+            InSemiOpenFromBelowInterval, ///< In semi-open from below interval
+            OneOf,                       ///< One of
+            NoneOf                       ///< None of
+        };
+
+        /// @brief Maps the comparison keys to valid range format string (used for the generation of error messages)
+        inline static std::unordered_map<Comparison, std::string> const ValidRangeFormat = {
+            {Comparison::Equal, "{} = {}"},
+            {Comparison::NotEqual, "{} != {}"},
+            {Comparison::Greater, "{} > {}"},
+            {Comparison::GreaterEqual, "{} >= {}"},
+            {Comparison::Less, "{} < {}"},
+            {Comparison::LessEqual, "{} <= {}"},
+            {Comparison::InClosedInterval, "{} <= {} <= {}"},
+            {Comparison::InOpenInterval, "{} <= {} <= {}"},
+            {Comparison::InSemiOpenFromAboveInterval, "{} <= {} < {}"},
+            {Comparison::InSemiOpenFromBelowInterval, "{} < {} <= {}"},
```

Comment:
(Not necessary to do)
Perhaps change name to InRightHalfOpenInterval and inLeftHalfOpenInterval

Reply:
Had a hard time choosing something suitable. Will change.

Reply:
Done